### PR TITLE
chore: add AthenaConfig to impl.py

### PR DESCRIPTION
**Pull request by SoumayaMauthoorMOJ**
### Description

Add `AthenaConfig` to `impl.py` following [recommendations](https://docs.getdbt.com/guides/dbt-ecosystem/adapter-development/3-building-a-new-adapter)

I have added the [Table configuration](https://github.com/dbt-athena/dbt-athena) as well as the connection parameters that can be passed to `cursor.execute()`

This might help to resolve https://github.com/dbt-athena/dbt-athena/issues/377, but it doesn't fix it by itself

## Models used to test - Optional
I ran the unit tests locally and they all passed
I ran some of the functional tests and they passed too, but I didn't run all of them

## Checklist
- [x] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [x] You kept your Pull Request small and focused on a single feature or bug fix.
- [x] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

File: dbt/adapters/athena/impl.py [hunk 1]
```diff
@@ -61,12 +61,32 @@
...
+    bucket_count: Optional[str] = None
+    table_type: Optional[str] = "hive"
+    ha: Optional[bool] = False
+    format: Optional[str] = "parquet"
+    write_compression: Optional[str] = None
+    field_delimiter: Optional[str] = None
+    table_properties: Optional[str] = None
+    native_drop: Optional[str] = None
+    seed_by_insert: Optional[bool] = False
+    lf_tags_config: Optional[str] = None
```
Comment (svdimchenko) [inline dbt/adapters/athena/impl.py:80]:
```suggestion
    lf_tags_config: Optional[Dict[str, Any]] = None
```

File: dbt/adapters/athena/impl.py [hunk 2]
```diff
@@ -61,12 +62,32 @@
 boto3_client_lock = Lock()
 
 
+@dataclass
+class AthenaConfig(AdapterConfig):
```
Comment (nicor88) [inline dbt/adapters/athena/impl.py:66]: nice 💯

**Comment by nicor88** [comment]
I believe that there is still some work to do no?
In https://github.com/dbt-athena/dbt-athena/issues/377 we decided to pick the workgroup from model config and pass it down to `adapter.run_query_with_partitions_limit_catching`, this means that in each macro you need to:

* pick the workgroup from config using jijia-> `{%- set work_group = config.get('work_group', default=target.work_group) -%}`
* inject workgroup when possible to when we execute queries

**Comment by SoumayaMauthoorMOJ** [comment]
Hey @nicor88 you're right there is still some work to do, but I thought I would start off with a smaller change first which might help with #377. In the meantime I've raised a question on slack channel https://getdbt.slack.com/archives/CBSQTAPLG/p1694451069966319 about the best way of tackling this. I'm worried about the amount of refactoring required if I modify all the materialisations.

**Comment by SoumayaMauthoorMOJ** [comment]
@nicor88 Did not know there was a slack channel raising now :-)

File: dbt/adapters/athena/impl.py (hunk 2 above)
Comment (svdimchenko) [reply dbt/adapters/athena/impl.py:67]: if I understand correct, this class should contain all additional model configs the adapter work with. If it's so, then we should add following configs:
```
seed_s3_upload_args: Otional[Dict[str, Any]] = None  # https://github.com/dbt-athena/dbt-athena/commit/9d029bbfd1fdab2c8c7f9ebe7185e7a34f4be63e
partitions_limit: Optinal[int] = None  #https://github.com/gr8-toolkit/dbt-athena/blob/4394dfa3abd20c8e87345de3f0a271582d56ebcb/dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql#L3-L3
```

Comment (svdimchenko) [reply dbt/adapters/athena/impl.py:67]: also It would be perfect to add docstring with these parameters description

Comment (nicor88) [reply dbt/adapters/athena/impl.py:67]: do we support partitions_limit globally? to me seems more a model config, e.g. [here](https://github.com/dbt-athena/dbt-athena/blob/9dec391257fce15b1887d8fa6bb8dd4b7e8261a2/dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql#L3)

Might not be a bad idea to do something like this
```
{%- set athena_partitions_limit = config.get('partitions_limit', target.partitions_limit) | int -%}
```

Comment (svdimchenko) [reply dbt/adapters/athena/impl.py:67]: > do we support partitions_limit globally?
now not, but such config can be added of course

**Review by svdimchenko — CHANGES_REQUESTED** [review]
the code looks nice, let's add missed parameters and docstring and we are ready to go

**Comment by SoumayaMauthoorMOJ** [comment]
@svdimchenko I don't have permission to add the label to run the functional tests, do think it's worthwhile running those? 

Also not too sure what docstring too used because explanation on [dbt website](https://docs.getdbt.com/guides/dbt-ecosystem/adapter-development/3-building-a-new-adapter) isn't very clear: database- and relation-level configs 

Hence why I raised a question on slack. I'll get the attribute descriptions from the README.

**Comment by svdimchenko** [comment]
> @svdimchenko I don't have permission to add the label to run the functional tests, do think it's worthwhile running those?
> 
> Also not too sure what docstring to use because explanation on [dbt website](https://docs.getdbt.com/guides/dbt-ecosystem/adapter-development/3-building-a-new-adapter) isn't very clear: "database- and relation-level configs"
> 
> Hence why I raised a question on slack. I'll get the attribute descriptions from the README.

I agree with you, docstring is not obligatory. However we should have all parameters defined I guess. So this comment should be fixed: https://github.com/dbt-athena/dbt-athena/pull/412#discussion_r1322474172

File: dbt/adapters/athena/impl.py [hunk 3]
```diff
@@ -80,8 +102,9 @@ class AthenaConfig(AdapterConfig):
     native_drop: Optional[str] = None
     seed_by_insert: Optional[bool] = False
     lf_tags_config: Optional[Dict[str, Any]] = None
-
-
+    seed_s3_upload_args: Optional[Dict[str, Any]] = None
+    partitions_limit: Optional[int] = None
```
Comment (nicor88) [inline dbt/adapters/athena/impl.py:106]: fine for now, as it's future proff - but that params is not used at all, we must add it as fallback in the model configs.

**Review by svdimchenko — APPROVED** [review]

**Review by svdimchenko — CHANGES_REQUESTED** [review]
@SoumayaMauthoorMOJ pls install and run pre-commit locally to avoid pre-commit checks fail in PR. Once it's fixed we are ready to go

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ it's recommended for new contributors to follow the [contributing](https://github.com/dbt-athena/dbt-athena/blob/main/CONTRIBUTING.md) section.
In there there is a step called `make setup` that also takes care of installing  the pre-commit hooks, therefore the same checks that we execute in the CI are executed before commiting - please have a look - exactly as @svdimchenko suggested btw.

**Comment by SoumayaMauthoorMOJ** [comment]
Sorry will do this evening  :-)

File: dbt/adapters/athena/impl.py [hunk 4]
```diff
@@ -61,12 +63,58 @@
 boto3_client_lock = Lock()
 
 
+@dataclass
+class AthenaConfig(AdapterConfig):
+    """This is .
```
Comment (nicor88) [inline dbt/adapters/athena/impl.py:68]: Aren't you missing something here?

**Review by nicor88 — APPROVED** [review]

**Review by svdimchenko — APPROVED** [review]

**Pull request merged**